kubernetes/apps_kubeclient.go

```diff
@@ -770,63 +867,433 @@ func (kc *kubeClient) getPods(namespace string, uid types.UID) ([]v1.Pod, error)
+		for _, status := range pod.Status.ContainerStatuses {
+			state := status.State
+			// Check if the container terminated with non-zero exit status
+			if state.Terminated != nil && state.Terminated.ExitCode != 0 {
+				// Severe if pod is terminated, indicating container didn't stop cleanly
+				return true, pod.DeletionTimestamp != nil
+			}
+			// Check if the container has been repeatedly crashing
+			if state.Waiting != nil && state.Waiting.Reason == containerCrashLoop {
+				return true, true
+			}
+			// Check if the container has not become ready within timeout
+			if state.Running != nil && !status.Ready {
+				startTime := state.Running.StartedAt.Time
+				duration := now.Sub(startTime)
+				if duration > containerTimeout {
+					return true, false
+				}
+			}
 		}
 	}
 
-	total := len(pods)
-	result := &app.PodStats{
-		Starting: &starting,
-		Running:  &running,
-		Stopping: &stopping,
-		Total:    &total,
+	return false, false
+}
+
+func isPullingImage(pod *v1.Pod) bool {
+	const containerCreating string = "ContainerCreating"
+	// If pod is pending with a container waiting due to a "ContainerCreating" event,
+	// categorize as "Pulling". This may change as more information is made available.
+	// See: https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/filters/resources.js#L663
+	if pod.Status.Phase == v1.PodPending {
+		for _, status := range pod.Status.ContainerStatuses {
+			waiting := status.State.Waiting
+			if waiting != nil && waiting.Reason == containerCreating {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isPodReady(pod *v1.Pod) bool {
+	// If all of the pod's containers have a ready status, then the pod is
+	// considered ready.
+	total := len(pod.Spec.Containers)
+	numReady := 0
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Ready {
+			numReady++
+		}
+	}
+	return numReady == total
+}
+
+func (kc *kubeClient) getBestRoute(namespace string, dc *deployment) (*url.URL, error) {
+	serviceMap, err := kc.getMatchingServices(namespace, dc)
+	if err != nil {
+		return nil, err
+	}
+	// Get routes and associate to services using spec.to.name
+	err = kc.getRoutesByService(namespace, serviceMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find route with highest score according to heuristics from web-console
+	var bestRoute *route
+	bestScore := -1
+	for _, routes := range serviceMap {
+		for _, route := range routes {
+			score := scoreRoute(route)
+			if score > bestScore {
+				bestScore = score
+				bestRoute = route
+			}
+		}
+	}
+
+	// Construct URL from best route
+	var result *url.URL
+	if bestRoute != nil {
+		scheme := "http"
+		if bestRoute.tls {
+			scheme = "https"
+		}
+		result = &url.URL{
+			Scheme: scheme,
+			Host:   bestRoute.host,
+			Path:   bestRoute.path,
+		}
 	}
 
 	return result, nil
 }
 
+func (kc *kubeClient) getMatchingServices(namespace string, dc *deployment) (routesByService map[string][]*route, err error) {
```

Comment:
is this method covered by 1 or more tests ?

Reply:
A whitebox test has not yet been written.